batch/batch/cloud/terra/azure/driver/driver.py

```diff
@@ -0,0 +1,554 @@
+      --mount type=bind,source=/host,target=/host \
+      --mount type=bind,source=/dev,target=/dev,bind-propagation=rshared \
+      -p 5000:5000 \
+      --device /dev/fuse \
+      --device /dev \
+      --privileged \
+      --cap-add SYS_ADMIN \
+      --security-opt apparmor:unconfined \
+      --network host \
+      $BATCH_WORKER_IMAGE \
+      python3 -u -m batch.worker.worker
+
+
+runcmd:
+  - nohup bash /startup.sh 2>&1 >worker.log &
+    '''
+
+    encoded_startup_script = base64.b64encode(startup_script.encode()).decode()
+
+    config = {
+        'common': {
+            'name': machine_name,
+            'description': machine_name,
+            'cloningInstructions': 'COPY_NOTHING',
+            'accessScope': 'PRIVATE_ACCESS',
+            'managedBy': 'USER',
+            'resourceId': instance_config._resource_id,
+            'properties': [],
+        },
+        'azureVm': {
+            'name': machine_name,
+            'vmSize': machine_type,
+            'vmImage': {
+                'publisher': 'Canonical',
+                'offer': '0001-com-ubuntu-server-focal',
+                'sku': '20_04-lts-gen2',
+                'version': '20.04.202305150',
+            },
+            'vmUser': {
+                'name': 'hail-admin',
+                'password': 'StrongP@ssword123',  # TODO Change
+            },
+            'ephemeralOSDisk': 'NONE',
+            'customData': encoded_startup_script,
+        },
+        'jobControl': {
+            'id': machine_name[32:],
+        },
+    }
+
+    if not instance_config.local_ssd_data_disk:
+        config['azureVm']['diskId'] = instance_config._disk_resource_id
+
+    return config
+
+
+class TerraAzureResourceManager(CloudResourceManager):
+    def __init__(
+        self,
+        billing_manager,
+    ):
+        self.terra_client = TerraClient()
+        self.billing_manager = billing_manager
+
+    async def delete_vm(self, instance: Instance):
+        config = instance.instance_config
+        assert isinstance(config, TerraAzureSlimInstanceConfig)
+        terra_vm_resource_id = config._resource_id
+
+        if not config.local_ssd_data_disk:
+            terra_disk_resource_id = config._disk_resource_id
+            await self.terra_client.post(
+                f'/disks/{terra_disk_resource_id}',
+                json={
+                    'jobControl': {'id': str(uuid.uuid4())},
+                },
+            )
+
+        try:
+            await self.terra_client.post(
+                f'/vm/{terra_vm_resource_id}',
+                json={
+                    'jobControl': {'id': str(uuid.uuid4())},
+                },
+            )
+        except aiohttp.ClientResponseError as e:
+            if e.status == 404:
+                raise VMDoesNotExist() from e
+            raise
+
+    async def get_vm_state(self, instance: Instance) -> VMState:
+        # TODO This should look at the response and use all applicable lifecycle types
```

Comment:
say more?

Reply:
Lost track of this in all the LoC in this change. In the [WSM API](https://workspace.dsde-dev.broadinstitute.org/swagger-ui.html#/ControlledAzureResource/getAzureVm) there's a `state` field in the response of the following enum `[ BROKEN, CREATING, DELETING, READY, UPDATING ]`. I'm hoping, but am not certain, that we can at least map `READY` to our notion of ready. Interestingly, this functionality doesn't seem entirely necessary, as a VM is marked ready regardless when it comes online and contacts the driver.